src/entry.c

```diff
@@ -72,109 +82,123 @@ sds entryGetField(const entry *entry) {
 
 /* Returns the location of a pointer to a separately allocated value. Only for
  * an entry without an embedded value. */
-static sds *entryGetValueRef(const entry *entry) {
+static inline void **entryGetValueRef(const entry *entry) {
     serverAssert(entryHasValuePtr(entry));
     char *field_data = sdsAllocPtr(entry);
-    field_data -= sizeof(sds);
-    return (sds *)field_data;
+    field_data -= sizeof(void *);
+    return (void **)field_data;
 }
 
-/* Returns the sds of the entry's value. */
-sds entryGetValue(const entry *entry) {
-    if (entryHasValuePtr(entry)) {
-        return *entryGetValueRef(entry);
-    } else {
+/* Returns the entry's value. */
+char *entryGetValue(const entry *entry, size_t *len) {
+    if (entryHasEmbeddedValue(entry)) {
         /* Skip field content, field null terminator and value sds8 hdr. */
         size_t offset = sdslen(entry) + 1 + sdsHdrSize(SDS_TYPE_8);
-        return (char *)entry + offset;
+        sds value = (char *)entry + offset;
+        if (len) *len = sdslen(value);
+        return value;
+    }
+    void **value_ref = entryGetValueRef(entry);
+    if (!value_ref) return NULL;
```

Comment:
Shouldn't this be an assertion?  We've already checked that the value is not embedded.  Is there a valid case for the reference to be NULL?  Note that none of the calling code is expecting this function to return NULL.  It's better to crash here rather than leaving the caller to crash.